### PR TITLE
[python] Refine pyenv based virtualenv management

### PR DESCRIPTION
**Pull request by indrajitr**
### Proposed Changes

Now that pyenv plugins availability is detected by directly probing
'pyenv', we need to rely on the same mechanism consistently.

Further, we perform available plugin scan with zsh native techniques
instead of relying on external commands.

Followup to #1433

### CODE REVIEW

**Comment by belak** [comment]
Quality as always. Looks good to me at a glance @indajitr.

File: modules/python/init.zsh [hunk 1]
```diff
@@ -92,22 +92,31 @@ if (( $+VIRTUALENVWRAPPER_VIRTUALENV || $+commands[virtualenv] )) && \
...
+    unset virtenv_plugins
   else
     # Fallback to 'virtualenvwrapper' without 'pyenv' wrapper in '$path'
     # and other known locations on a Debian based system.
     virtenv_sources=(
       ${(@Ov)commands[(I)virtualenvwrapper(_lazy|).sh]}
       /usr/share/virtualenvwrapper/virtualenvwrapper(_lazy|).sh(OnN)
     )
+
     source "${virtenv_sources[1]}"
```
Comment (zachwhaley) [inline modules/python/init.zsh]: FYI, this gets called whether this virtualenvwrapper file exists or not. 

So if you have pyenv installed, but not the virtualenv plugin or the virtualenvwrapper plugin, then you get a"file not found" error at login

File: modules/python/init.zsh [hunk 2]
```diff
@@ -92,22 +92,31 @@ if (( $+VIRTUALENVWRAPPER_VIRTUALENV || $+commands[virtualenv] )) && \
...
 
-  # Enable 'virtualenv' with 'pyenv'.
-  if (( $+commands[pyenv] )) && \
-     pyenv commands | command grep -q virtualenv-init
-  then
-    eval "$(pyenv virtualenv-init -)"
-    # Optionall activate 'virtualenvwrapper' with 'pyenv' is available.
-    if (( $#commands[(i)pyenv-virtualenvwrapper(_lazy|)] )); then
-      pyenv "${${(@O)commands[(I)pyenv-virtualenvwrapper(_lazy|)]}[1]#pyenv-}"
+  if (( $+commands[pyenv] )); then
```
Comment (zachwhaley) [inline modules/python/init.zsh]: I think this might be a regression for @diraol, since this might revert the changes in #1414

File: modules/python/init.zsh [hunk 3]
```diff
@@ -92,22 +92,31 @@ if (( $+VIRTUALENVWRAPPER_VIRTUALENV || $+commands[virtualenv] )) && \
...
+    # Grab sorted array of available 'pyenv' plugins by probing 'pyenv'
```
Comment (zachwhaley) [inline modules/python/init.zsh]: I'm not sure this comment matches the code pre se.  It looks like the code is returning all pyenv commands that start with "virtualenv", and not all pyenv plugins, correct?

File: modules/python/init.zsh [hunk 4]
```diff
@@ -92,18 +92,26 @@ if (( $+VIRTUALENVWRAPPER_VIRTUALENV || $+commands[virtualenv] )) && \
...
+  # Create a sorted array of available virtualenv related 'pyenv' commands to
+  # look for plugins of interest. Scanning shell '$path' isn't enough as they
+  # can exist in 'pyenv' synthesized paths (e.g., '~/.pyenv/plugins') instead.
+  local -a virtenv_plugins
+  if (( $+commands[pyenv] )); then
+    virtenv_plugins=(${(@oM)${(f)"$(pyenv commands --no-sh 2>/dev/null)"}:#virtualenv*})
+  fi
+
+  if (( $+commands[pyenv] && \
+        $virtenv_plugins[(i)virtualenv-init] <= $#virtenv_plugins )); then
```
Comment (diraol) [inline modules/python/init.zsh]: @indrajitr if `virtenv_plugins` is "empty", this expression (`$virtenv_plugins[(i)virtualenv-init] <= $#virtenv_plugins`) returns False?

I'm asking it because if there is no `pyenv` installed, then `virtenv_plugins` is at its initial state (`local -a virtenv_plugins`). Then, may be we do not need to re-check for `$+commands[pyenv]` here right?

File: modules/python/init.zsh (hunk 1 above)
Comment (indrajitr) [inline modules/python/init.zsh]: Good find, thanks! Applied the fix separately.

File: modules/python/init.zsh (hunk 2 above)
Comment (indrajitr) [inline modules/python/init.zsh]: Right you are! Clearly, I wasn't in right mind. Thanks.

File: modules/python/init.zsh (hunk 3 above)
Comment (indrajitr) [inline modules/python/init.zsh]: Reworded, does the new one make sense?

**Comment by indrajitr** [comment]
Thanks for the review guys!
